src/man/man1/owfs.man
'\"
'\" Copyright (c) 2003-2004 Paul H Alfille, MD
'\" (paul.alfille@gmail.com)
'\"
'\" Program manual page for the OWFS -- 1-wire filesystem package
'\" Based on Dallas Semiconductor, Inc's datasheets, and trial and error.
'\"
'\" Free for all use. No warranty. None. Use at your own risk.
'\"
.TH OWFS 1 2004 "OWFS Manpage" "One-Wire File System"
.SH NAME
.B owfs
\- 1-wire filesystem
.SH SYNOPSIS
.B owfs
.so man1/cmdline_mini.1so
.I \-m mountdir
.SH "DESCRIPTION"
.so man1/description.1so
.SS owfs
.B owfs (1)
is the filesystem client of the 
.I OWFS
family of programs. It runs on linux, freebsd and Mac OS X, and requires the 
.I fuse
kernel module and library. (http://fuse.sourceforge.net) which is a user-mode filesystem driver.
.PP
Essentially, the entire 1-wire bus is mounted to a place in your filesystem. All the 1-wire devices are accessible using standard file operations (read, write, directory listing). The system is safe, no actual files are exposed, these files are virtual. Not all operations are supported. Specifically, file creation, deletion, linking and renaming are not allowed. (You can link from outside to a owfs file, but not the other way around).
.so man1/device.1so
.SH SPECIFIC OPTIONS
.SS \-m \-\-mountpoint=directory_path
Path of a directory to mount the 1-wire file system
.PP
The mountpoint is required. There is no default.
.SS \-\-allow_other
Shorthand for fuse mount option "\-o allow_other"  Allows uther users to see the fuse (owfs) mount point and file system. Requires a setting in /etc/fuse.conf as well.
.SS \-\-fuse-opt "options"
Sends options to the fuse-mount process. Options should be quoted, e.g. "\"\-o allow_other\"" .
.so man1/temperature.1so
.so man1/pressure.1so
.so man1/format.1so
.so man1/job_control.1so
.so man1/configuration.1so
+.PP
+On Debian,
+.B owfs (1)
+always reads
+.I /etc/owfs.conf
.so man1/help.1so
.so man1/timeout.1so
.SH EXAMPLE
.TP
owfs \-d /dev/ttyS0 \-m /mnt/1wire
Bus master on serial port
.TP
owfs \-F \-u \-m /mnt/1wire
USB adapter, temperatures reported in Fahrenheit
.TP
owfs \-s 10.0.1.2:4304 \-m /mnt/1wire
Connect to an 
.B owserver (1)
process that was started on another machine at tcp port 4304
.SH SEE ALSO
.so man1/seealso.1so
.SH AVAILABILITY
http://www.owfs.org
.SH AUTHOR
Paul Alfille (paul.alfille@gmail.com)


debian/changelog
owfs (3.2p3+smurf-4) unstable; urgency=medium

  * *Grumble*

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 04 Jul 2020 13:40:13 +0200

owfs (3.2p3+smurf-3) unstable; urgency=medium

  * Fix rules

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 04 Jul 2020 13:35:02 +0200

owfs (3.2p3+smurf-2) unstable; urgency=medium

  * Use py3

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 04 Jul 2020 13:33:42 +0200

owfs (3.2p3+smurf-1) unstable; urgency=medium

  * Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 04 Jul 2020 13:30:40 +0200

owfs (3.2p3+release-2) unstable; urgency=medium

  * php and tcl are gone

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 20 Jun 2019 17:51:44 +0200

owfs (3.2p3+release-1) unstable; urgency=medium

  * Merge release

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 20 Jun 2019 17:39:58 +0200

owfs (3.2p3+git1-ba66022-4) unstable; urgency=medium

  * lots of README nonsense
  * .gitignore

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 20 Sep 2018 20:48:19 +0200

owfs (3.2p3+git1-ba66022-1) unstable; urgency=medium

  * Upstream Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 20 Sep 2018 18:27:24 +0200

owfs (3.1p3+git1-b7e91db-1) smurf; urgency=medium

  * Upstream Merge

 -- Matthias Urlichs <matthias@urlichs.de>  Wed, 17 Aug 2016 06:28:30 +0200

owfs (3.1p0+git1-ba7b185-4) smurf; urgency=medium

  * Fix dependencies
  * Install systemd unit files

 -- Matthias Urlichs <matthias@urlichs.de>  Tue, 16 Jun 2015 07:40:04 +0200

owfs (3.1p0+git1-ba7b185-3) smurf; urgency=medium

  * Status fix

 -- Matthias Urlichs <matthias@urlichs.de>  Sat, 06 Jun 2015 19:38:16 +0200

owfs (3.1p0+git1-ba7b185-1) unstable; urgency=medium

  * Don't forget to tag your releases!

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 04 Jun 2015 20:18:23 +0200

owfs (3.0p0+git1-ba7b185-2) unstable; urgency=medium

  * Merge to current git

 -- Matthias Urlichs <matthias@urlichs.de>  Thu, 04 Jun 2015 12:09:35 +0200

owfs (3.0p0-2) unstable; urgency=medium

  * include MoaT device

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 31 May 2015 10:04:58 +0200

owfs (3.0p0-1) unstable; urgency=medium

  * Build version 3.0p0

 -- Matthias Urlichs <matthias@urlichs.de>  Sun, 08 Mar 2015 11:56:58 +0100

owfs (2.9p8-6) unstable; urgency=medium

  * Hide ls error messages when nothing have to do done on
    upgrade (Closes: #769523)

 -- Vincent Danjean <vdanjean@debian.org>  Fri, 14 Nov 2014 18:54:21 +0100

owfs (2.9p8-5) unstable; urgency=medium

  * Add missing Breaks/Replaces required due to some files moved
    from the ow-shell package to the ow-tools one. This move occures
    in 2.9p5-1 (Closes: #768115)

 -- Vincent Danjean <vdanjean@debian.org>  Sat, 08 Nov 2014 23:09:09 +0100

owfs (2.9p8-4) unstable; urgency=medium

  * Fix "FTBFS[kfreebsd]: wrong ifdef __FreeBSD__" : using the good
    symbol, thanks Steven Chamberlain (Closes: #766910)

 -- Vincent Danjean <vdanjean@debian.org>  Mon, 27 Oct 2014 08:18:46 +0100

owfs (2.9p8-3) unstable; urgency=medium

  * Trying to fix another incompatibility with FreeBSD with this
    new upstream version

 -- Vincent Danjean <vdanjean@debian.org>  Sun, 26 Oct 2014 16:31:25 +0100

owfs (2.9p8-2) unstable; urgency=medium

  * make systemd dependency linux-any only (Closes: #766697)

 -- Vincent Danjean <vdanjean@debian.org>  Sat, 25 Oct 2014 10:36:53 +0200

owfs (2.9p8-1) unstable; urgency=medium

  * New upstream release
    + fix a nasty bug in owfs (filesystem)
    + will require a new round in NEW queue (sorry FTP Masters)
      as upstream uses a new soname for each release
      (by chance, there is no reverse dependencies in Debian, so
      no bin-NMU are required)

 -- Vincent Danjean <vdanjean@debian.org>  Sun, 19 Oct 2014 13:56:13 +0200

owfs (2.9p7-2) unstable; urgency=medium

  * remove README.source template
  * fix perl source file licenses in copyright and reorganize it
    (Closes: #765665)

 -- Vincent Danjean <vdanjean@debian.org>  Fri, 17 Oct 2014 15:08:15 +0200

owfs (2.9p7-1) unstable; urgency=medium

  * New (minor) upstream release
    + fix version parsing in python and perl modules (Closes: #765371)

 -- Vincent Danjean <vdanjean@debian.org>  Tue, 14 Oct 2014 18:57:13 +0200

owfs (2.9p5-1.1) unstable; urgency=high

  * Non-maintainer upload with maintainer's permission

  * make DS1410_detect@Base symbol available only on Linux architectures.
    Fixes FTBFS on kfreebsd-*/hurd (Closes: #757571)

  * High urgency to help the perl 5.20 transition

 -- Damyan Ivanov <dmn@debian.org>  Tue, 19 Aug 2014 08:30:50 +0000

owfs (2.9p5-1) unstable; urgency=medium

  * New upstream release
  * Fix "hardcodes /usr/lib/perl5":
    make debian/libow-perl.install executable and use $Config{vendorarch}
    there (Closes: #752804)
    Thanks Gregor Herrmann <gregoa@debian.org> for the patch

 -- Vincent Danjean <vdanjean@debian.org>  Mon, 07 Jul 2014 13:27:43 +0200

owfs (2.9p1-2) unstable; urgency=medium

  * move TclTk scripts to an arch-indep package (Closes: #737400)
  * upload to unstable
  * Bump Standard-Version (no change needed)
  * Fix URLs in debian/control
  * add a Suggests on avahi-daemon in ow-server and document the --nozero
    option (closes: #646764)
  * recommands fuse instead of fuse-utils (Closes: #698180)

 -- Vincent Danjean <vdanjean@debian.org>  Sun, 02 Feb 2014 22:07:24 +0100

owfs (2.9p1-1) experimental; urgency=low

  * Imported Upstream version 2.9p1
  * Cleanup patches and symbol files

 -- Vincent Danjean <vdanjean@debian.org>  Wed, 26 Jun 2013 23:52:10 +0200

owfs (2.9p0-1) experimental; urgency=low

  * Imported Upstream version 2.9p0
  * Update to new release 2.9-0
  * backport python fix in ownet

 -- Vincent Danjean <vdanjean@debian.org>  Tue, 02 Apr 2013 21:19:59 +0200

owfs (2.8p17-1) experimental; urgency=low

  * New upstream release

 -- Vincent Danjean <vdanjean@debian.org>  Sun, 09 Sep 2012 22:04:33 +0200

owfs (2.8p16-1) unstable; urgency=low

  * New upstream release


module/owfs/src/c/owfs.c
/*
     OW -- One-Wire filesystem
    version 0.4 7/2/2003

    Function naming scheme:
    OW -- Generic call to interaface
    LI -- LINK commands
    L1 -- 2480B commands
    FS -- filesystem commands
    UT -- utility functions
    COM - serial port functions
    DS2480 -- DS9097 serial connector

    Written 2003 Paul H Alfille
*/

#include "owfs.h"
#include "ow_connection.h"
#ifdef __FreeBSD_kernel__
#include "sys/param.h"
#endif
#include "sys/mount.h"

/*
    OW -- One Wire
    Globals variables -- each invokation will have it's own data
*/
struct fuse *fuse;
int fuse_fd = -1;
char umount_cmd[1024] = "";
char *fuse_mnt_opt = NULL;
char *fuse_open_opt = NULL;

/* ---------------------------------------------- */
/* Command line parsing and mount handler to FUSE */
/* ---------------------------------------------- */
int main(int argc, char *argv[])
{
	int c;
	struct Fuse_option fuse_options;

	/* Set up owlib */
	LibSetup(program_type_filesystem);

	/* grab our executable name */
	ArgCopy( argc, argv ) ;

	//mtrace() ;
	/* process command line arguments */
	while ((c = getopt_long(argc, argv, OWLIB_OPT, owopts_long, NULL)) != -1) {
		switch (c) {
		case 'V':
			fprintf(stderr, "%s version:\n\t" VERSION "\n", argv[0]);
			break;
		case e_fuse_opt:		/* fuse_mnt_opt */
			if (fuse_mnt_opt) {
				// delete prior
				owfree(fuse_mnt_opt);
			}
			fuse_mnt_opt = Fuse_arg(optarg, "FUSE mount options");
			if (fuse_mnt_opt == NULL) {
				ow_exit(0);
			}
			break;
		case e_fuse_open_opt:	/* fuse_open_opt */
			if (fuse_open_opt) {
				// delete prior
				owfree(fuse_open_opt);
			}
			fuse_open_opt = Fuse_arg(optarg, "FUSE open options");
			if (fuse_open_opt == NULL) {
				ow_exit(0);
			}
			break;
		}
		// rest of options -- use common opt arguments code
		if ( BAD( owopt(c, optarg) ) ) {
			ow_exit(0);	
		}
	}

	/* non-option arguments */
	while (optind < argc - 1) {
		ARG_Generic(argv[optind]);
		++optind;
	}
	while (optind < argc) {
		if (Outbound_Control.active) {
			ARG_Generic(argv[optind]);
		} else {
			ARG_Server(argv[optind]);
		}
		++optind;
	}

	if (Outbound_Control.active == 0) {
		LEVEL_DEFAULT("No mount point specified.\nTry '%s -h' for help.", argv[0]);
		ow_exit(1);
	}
	// FUSE directory mounting
	LEVEL_CONNECT("fuse mount point: %s", Outbound_Control.head->name);

	// Signal handler is set in fuse library
	//set_signal_handlers(exit_handler);

	/* Set up adapters */
	if ( BAD(LibStart(NULL)) ) {
		ow_exit(1);
	}

#if FUSE_VERSION >= 14
	/* Set up "command line" for main fuse routines */
	Fuse_setup(&fuse_options);	// command line setup
	Fuse_add(Outbound_Control.head->name, &fuse_options);	// mount point
#if FUSE_VERSION >= 22
	Fuse_add("-o", &fuse_options);	// add "-o direct_io" to prevent buffering
	Fuse_add("direct_io", &fuse_options);
#endif							/* FUSE_VERSION >= 22 */
	switch (Globals.daemon_status) {
		case e_daemon_fg:
			Fuse_add("-f", &fuse_options);	// foreground for fuse too
			if (Globals.error_level > 2) {
				Fuse_add("-d", &fuse_options);	// debug for fuse too
			}
//			Globals.daemon_status = e_daemon_bg;
			break ;
		default:
			break ;
	}		

	// Unmount just in case
	// No checks -- can fail without consequences
#if defined(__FreeBSD__) || defined(__APPLE__)
	unmount( Outbound_Control.head->name, 0) ;
#else
	umount( Outbound_Control.head->name ) ;
#endif

	Fuse_parse(fuse_mnt_opt, &fuse_options);
	LEVEL_DEBUG("fuse_mnt_opt=[%s]", fuse_mnt_opt);
	Fuse_parse(fuse_open_opt, &fuse_options);
	LEVEL_DEBUG("fuse_open_opt=[%s]", fuse_open_opt);
	if (Globals.allow_other) {
		Fuse_add("-o", &fuse_options);	// add "-o allow_other" to permit other users access
		Fuse_add("allow_other", &fuse_options);
	}


#if FUSE_VERSION > 25
	fuse_main(fuse_options.argc, fuse_options.argv, &owfs_oper, NULL);
#else							/* FUSE_VERSION <= 25 */
	fuse_main(fuse_options.argc, fuse_options.argv, &owfs_oper);
#endif							/* FUSE_VERSION <= 25 */
	Fuse_cleanup(&fuse_options);
#endif

	ow_exit(0);
	return 0;
}
